kpi/views/v2/asset.py

```diff
@@ -203,6 +210,37 @@ def get_queryset(self, *args, **kwargs):
             # have to care about optimizations for that?
             return queryset
 
+    def get_serializer_context(self):
+        """
+        Extra context provided to the serializer class.
+        """
+
+        context_ = super(AssetViewSet, self).get_serializer_context()
+        if self.action == 'list':
+            # To avoid making a triple join-query for each asset in the list
+            # to retrieve object permissions. We make a big one at beginning
+            # to build an dict key-ed by asset ids.
+            # The serializer will be able to pick what it needs from that dict
+            # and narrow down data according to users' permissions.
+            queryset = super(AssetViewSet, self).get_queryset()
+            asset_content_type = ContentType.objects.get_for_model(Asset)
+            asset_ids = self.filter_queryset(queryset).values_list('id').distinct()
+            object_permissions = ObjectPermission.objects.\
+                filter(content_type_id=asset_content_type.pk,
+                       object_id__in=asset_ids).\
+                select_related('user', 'permission')
+
+            object_permissions_per_object = defaultdict(list)
+
+            for op in object_permissions:
+                object_permissions_per_object[op.object_id].append(op)
+
+            context_.update({
+                'object_permissions_per_object': object_permissions_per_object
+            })
```

Comment:
A quick glance at DRF makes me think `context_` is a `dict`; could we `context_['object_permissions_per_object'] = object_permissions_per_object`?